apps/aecore/src/aec_tx_pool.erl

```diff
@@ -628,13 +661,110 @@ update_pool_on_tx_hash(TxHash, {#dbs{gc_db = GCDb} = Dbs, GCHeight}, Handled) ->
                     %% Added to chain
                     Key = pool_db_key(Tx),
                     pool_db_raw_delete(Dbs, Key),
-                    aec_tx_pool_gc:delete_hash(GCDb, TxHash);
+                    aec_tx_pool_gc:delete_hash(GCDb, TxHash),
+                    ok = add_to_origins_cache(OriginsCache, Tx);
                 true ->
                     Key = pool_db_key(Tx),
                     pool_db_raw_put(Dbs, GCHeight, Key, Tx, TxHash)
             end
     end.
 
+%%===================================================================
+%% Origins cache
+%%
+%% Origins/accounts cache keeps track of {origin, nonce} pairs which
+%% have been added to the chain. This allows earlier
+%% garbage collection of potentially hanging transactions from
+%% the same origin/account, but with nonce that will block them
+%% from ever getting into the chain.
+%% E.g. if a transaction from sender X with nonce Y made it into the
+%% chain, all pending transactions from X with nonce =< Y can
+%% be removed from the mempool.
+
+add_to_origins_cache(OriginsCache, SignedTx) ->
+    case ets:info(OriginsCache, size) < ?ORIGINS_CACHE_MAX_SIZE of
+        true ->
+            Tx     = aetx_sign:tx(SignedTx),
+            Origin = aetx:origin(Tx),
+            Nonce  = aetx:nonce(Tx),
+            case ets:lookup(OriginsCache, Origin) of
+                [] ->
+                    true = ets:insert(OriginsCache, {Origin, Nonce}),
+                    ok;
+                [{Origin, N}] ->
+                    %% Always keep the highest included nonce in cache
+                    case Nonce > N of
+                        true ->
+                            true = ets:insert(OriginsCache, {Origin, Nonce}),
+                            ok;
+                        false -> ok
+                    end
+            end;
+        false -> ok
+    end.
+
+origins_cache_gc(#state{origins_cache = OriginsCache, dbs = Dbs}) ->
+    EntriesToProcess = 100,
```

Comment:
100 is a number from a hat? Might be good to indicate that...

Reply:
Yes, I'll put it into the constant to tune it if needed.